### ISSUE TITLE
G_KEEPERR clearing ERRSV in newer perl

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12664]
Migrated from [rt.perl.org#116170](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=116170) (status was 'open')

Searchable as RT116170$

**Comment by p5pRT** [comment]
### From jims@conceptual.com.au

#### Created by jims@conceptual.com.au

This is a bug report for perl from jims@&#8203;conceptual\.com\.au\,
generated with the help of perlbug 1\.39 running under perl 5\.14\.2\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
G\_KEEPERR when set in embedded perl in c\, causes ERRSV to be cleared
when calling die from within the perl code\.

This is a new issue\, it works as expected on older perls like 5\.10\.1\.\.

$ cat >a\.c
\#include \<stdio\.h>

\#include \<EXTERN\.h> /\* from the Perl distribution \*/
\#include \<perl\.h> /\* from the Perl distribution \*/
static PerlInterpreter \*my\_perl; /\*\*\* The Perl interpreter \*\*\*/
int main\(int argc\, char \*\*argv\, char \*\*env\) \{
&nbsp;    PERL\_SYS\_INIT3\(&argc\,&argv\,&env\);
&nbsp;    my\_perl = perl\_alloc\(\);
&nbsp;    perl\_construct\(my\_perl\);
&nbsp;    PL\_exit\_flags |= PERL\_EXIT\_DESTRUCT\_END;
&nbsp;    perl\_parse\(my\_perl\, NULL\, argc\, argv\, \(char \*\*\)NULL\);

&nbsp;    SV \*func = newSVpv\("function"\, 0\);
&nbsp;    int res = call\_sv\(func\, G\_EVAL | G\_KEEPERR\);

&nbsp;    if\( SvTRUE\(ERRSV\) \) \{
&nbsp;        printf\("died\\n"\);
&nbsp;    \} else \{
&nbsp;        printf\("good exit\\n"\);
&nbsp;    \}
&nbsp;    
&nbsp;    perl\_destruct\(my\_perl\);
&nbsp;    perl\_free\(my\_perl\);
&nbsp;    PERL\_SYS\_TERM\(\);
\}
CTRL\-D

$ gcc a\.c \-o a \`perl \-MExtUtils&#8203;::Embed \-e ccopts \-e ldopts\`
$ \./a \-e 'sub function\(\) \{ die "please" \}'
didn't die



On a perl 5\.10\.1 machine I get the expected result "died"\.

Also removing '| G\_KEEPERR' also produced the expected result "died"\.


Cheers

James Shirley

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.14.2:

Configured by Debian Project at Tue Nov 27 00:34:23 UTC 2012.

Summary of my perl5 (revision 5 version 14 subversion 2) configuration:

  Platform:
    osname=linux, osvers=2.6.42-32-generic,
archname=x86_64-linux-gnu-thread-multi
    uname='linux allspice 2.6.42-32-generic #51-ubuntu smp wed sep 26
21:33:09 utc 2012 x86_64 x86_64 x86_64 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN
-Dcccdlflags=-fPIC -Darchname=x86_64-linux-gnu -Dprefix=/usr
-Dprivlib=/usr/share/perl/5.14 -Darchlib=/usr/lib/perl/5.14
-Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5
-Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local
-Dsitelib=/usr/local/share/perl/5.14.2
-Dsitearch=/usr/local/lib/perl/5.14.2 -Dman1dir=/usr/share/man/man1
-Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1
-Dsiteman3dir=/usr/local/man/man3 -Duse64bitint -Dman1ext=1
-Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh
-Ud_ualarm -Uusesfio -Uusenm -Ui_libutil -DDEBUGGING=-g -Doptimize=-O2
-Duseshrplib -Dlibperl=libperl.so.5.14.2 -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include
-D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing
-pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.6.3', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/x86_64-linux-gnu /lib/../lib
/usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=, so=so, useshrplib=true, libperl=libperl.so.5.14.2
    gnulibc_version='2.15'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -L/usr/local/lib
-fstack-protector'

Locally applied patches:



@INC for perl 5.14.2:
    /etc/perl
    /usr/local/lib/perl/5.14.2
    /usr/local/share/perl/5.14.2
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.14
    /usr/share/perl/5.14
    /usr/local/lib/site_perl
    .


Environment for perl 5.14.2:
    HOME=/home/jims
    LANG=en_AU.UTF-8
    LANGUAGE=en_AU:en
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/lib/lightdm/lightdm:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @bulk88

On Fri Dec 21 23&#8203;:13&#8203;:22 2012\, jims@&#8203;conceptual\.com\.au wrote&#8203;:
> This is a bug report for perl from jims@&#8203;conceptual\.com\.au\,
> generated with the help of perlbug 1\.39 running under perl 5\.14\.2\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> G\_KEEPERR when set in embedded perl in c\, causes ERRSV to be cleared
> when calling die from within the perl code\.
> 
> This is a new issue\, it works as expected on older perls like 5\.10\.1\.\.
> 
> $ cat >a\.c
> \#include \<stdio\.h>
> 
> \#include \<EXTERN\.h> /\* from the Perl distribution \*/
> \#include \<perl\.h> /\* from the Perl distribution \*/
> static PerlInterpreter \*my\_perl; /\*\*\* The Perl interpreter \*\*\*/
> int main\(int argc\, char \*\*argv\, char \*\*env\) \{
>     PERL\_SYS\_INIT3\(&argc\,&argv\,&env\);
>     my\_perl = perl\_alloc\(\);
>     perl\_construct\(my\_perl\);
>     PL\_exit\_flags |= PERL\_EXIT\_DESTRUCT\_END;
>     perl\_parse\(my\_perl\, NULL\, argc\, argv\, \(char \*\*\)NULL\);
> 
>     SV \*func = newSVpv\("function"\, 0\);
>     int res = call\_sv\(func\, G\_EVAL | G\_KEEPERR\);
> 
>     if\( SvTRUE\(ERRSV\) \) \{
>         printf\("died\\n"\);
>     \} else \{
>         printf\("good exit\\n"\);
>     \}
> 
>     perl\_destruct\(my\_perl\);
>     perl\_free\(my\_perl\);
>     PERL\_SYS\_TERM\(\);
> \}
> CTRL\-D
> 
> $ gcc a\.c \-o a \`perl \-MExtUtils&#8203;::Embed \-e ccopts \-e ldopts\`
> $ \./a \-e 'sub function\(\) \{ die "please" \}'
> didn't die
> 
> 
> 
> On a perl 5\.10\.1 machine I get the expected result "died"\.
> 
> Also removing '| G\_KEEPERR' also produced the expected result "died"\.
> 
> 
> Cheers
> 
> James Shirley
> 

Dont know if this is related or not\, ERRSV should never be placed into a
multiple evaluation macro like SvTRUE\. Assign ERRSV to a regular auto SV
\*\. Then do SvTRUE\. Or use sv\_true\(ERRSV\)\. Or use SvTRUEx \(not public
api\) which evaluates only once\.

\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'
